MSC3489: Sharing streams of location data with history

ara4n opened the pull request:
In which we introduce `m.beacon_info.*` and `m.beacon.*` events for streaming location data, building on [MSC3488](https://github.com/matrix-org/matrix-doc/pull/3488).

[Rendered](https://github.com/matrix-org/matrix-doc/blob/matthew/location-streaming/proposals/3489-location-streaming.md)

**Matrix React SDK implementation**

https://github.com/matrix-org/matrix-react-sdk/tree/develop/src/components/views/location

**Element Android implementation**

https://github.com/vector-im/element-android/tree/develop/vector/src/main/java/im/vector/app/features/location

**Matrix iOS SDK implementation**

https://github.com/matrix-org/matrix-ios-sdk/tree/develop/MatrixSDK/LocationSharing






















<!-- Replace -->
Preview: https://pr3489--matrix-org-previews.netlify.app
<!-- Replace -->

ara4n committed:
MSC3489: Sharing streams of location data

ara4n committed:
clarify description param

ara4n committed:
json5

ara4n committed:
bullets

ara4n committed:
typos

vranki commented:
Looks good, but I don't like this part: "m.beacon.* events should be sent every 2 seconds while the location of the asset is moving. If the asset is not moving, it should be refreshed every 30 seconds."

The location data can come from another source via bridge or bot and this data source may have it's own update interval. Also the practical update rate depends on application. Some applications may need 2 seconds, but some may be happy with 60 secs or even longer. An application may choose use smart algorithm that calculates error between true location and location extrapolated from velocity vector and updates only if it grows high enough.

I'd change this to recommendation, although 2 secs is quite high rate, even for moving cars or people being followed live on map. Maybe there should also be a field for maximum time of next update. If the update isn't sent before this, the recipient can assume that sender was unable to send update for some reason.

ara4n committed:
incorporate feedback

relax rate requirements
spell out that we're ignoring velocity for now

ara4n commented:
feedback incorporated - thank you. (please always provide feedback by commenting on the relevant line of the PR, or the heading if there isn't a specific relevant line, so that we can track the feedback & resolve it as a thread, until github *finally* implements threaded comments...)

ara4n commented:
feedback incorporated - thank you. (please always provide feedback by commenting on the relevant line of the PR, or the heading if there isn't a specific relevant line, so that we can track the feedback & resolve it as a thread, until github *finally* implements threaded comments...)

ananace commented:
Sneaking in a little note here that another alternative could be done with a combination of MSC2477 and a slower rate of state events as in here for the "breadcrumb trail" - even described that particular use-case as an example in the proposal there. Something much like this (like Telegram's live location) was one of the original thoughts for a use-case when I first wrote that proposal after all.

ara4n commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+behaviour we need (easy access to current value; persistent state within
+a room), hence adopting them for this.
+
+## Alternatives
```
Need to add @ananace's sugggestion of using EDUs from https://github.com/matrix-org/matrix-doc/pull/2477 to fan out the location data to the room.  The main downside I see is that you lose persistence (but could send the same events as non-EDUs for persistence) - and E2EE EDUs don't exist yet.

ananace replied:
I'm not all that well read into what's required for E2EE to function, but wouldn't it be possible to use the user-defined EDUs to send an EDU inside an encryption envelope, just like with regular encrypted PDUs?

Sorunome replied:
It should work the exact same way, yes

andybalaam replied:
EDU-related stuff is now all in https://github.com/matrix-org/matrix-spec-proposals/pull/3672

PaarthShah commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+say what the various beacons are doing rather than simply fishing it out of
+room state.
+
+Alternatively, we could negotiate a WebRTC data channel using MSC3401 and
+stream low-latency geospatial data over the participating devices in the room.
+This doesn't give us historical data, however, and requiring a WebRTC stack
+is prohibitively complicated for simple clients (e.g. basic IOT devices
+reporting spatial telemetry).
```
While I agree that WebRTC could possibly be overengineering the problem, I'd posit that historical data should not be a priority for this proposal at all. Many existing location-sharing functionalities (I'll take google maps, as an example) do not share a full record of the journey to people with whom the location was shared in the duration.

Recording the resulting full GPS path and being able to share it (most ideally in GPX format, though there are [alternatives](https://opengpxfile.com/alternatives-to-gpx.html)) could be an event unto itself; at its simplest, uploading the final resulting file when the location share concludes.

PaarthShah replied:
The event involved with sharing the resulting GPX file could be rendered client-side specially to display the path taken, perhaps even "playing back" the live location on an (accelerated?) timeline.

andybalaam replied:
This MSC has now been re-worded to make it exclusively about the with-history case. The no-history case is now in https://github.com/matrix-org/matrix-spec-proposals/pull/3672

deepbluev7 commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+   * means new joiners to a room won't even be able to see the most recent location for someone
+   * MSC1763 data retention lets us control the historical data anyway.
+
+Alternatively, we could send location data as normal E2EE messages. However,
+clients would have to back-paginate in an unbounded fashion to confidently
+say what the various beacons are doing rather than simply fishing it out of
+room state.
```
I prefer sending just normal location messages, that get edited. Pros:

- E2EE
- Automatic fallback, since it is just an edit
- No strain on state res

The downsides of being hard to fish out from the history should at least not matter for most IM usages. If there was a recent update, it is in the last few messages already. And if you implement them like most other messengers, you will be showing the live location as a message in the timeline, to which the edits are applied, so you don't need to do any fishing anyway.

And if fishing really is an issue, one could just introduce a state event like this:

```json5
{
  "type": "m.locations",
  "state_key": "@user_id:server.tld",
  "content": {
    "m.active_locations": [
      "$eventid1",
      "$eventid2"
    ]
  }
}
```

Which provides links to the root location event of the edit chain and would only need updating when starting or stopping the sharing. For other usecases one could also use a different state event key, the userid could just be the default for sharing your own location.

andybalaam replied:
The proposal (as recently updated) suggests one state event every time a user starts or stops sharing, and normal events to send the actual locations. I think this strikes a good compromise because:

* Clients are able to look up easily in room state who is sharing location and whether the share is live, and
* Locations are encrypted because they are not in the state event.

The downsides you mention are:

1. More state events in history, making state resolution slower
2. More state sent to clients

For 1, I see this as a legitimate use of room state: we want to know this when we enter a room, and don't want to have to search the timeline to see who is sharing.

I believe we should address 2 by allowing deleting state (probably by redacting state events) such that clients no longer see deleted state. I argue that deleting state is orthogonal to this proposal.

Your proposal for a single `m.locations` event only addresses point 2 I think. It looks like a plausible solution, but I don't see a great advantage over what is in the MSC. Could you explain further?

deepbluev7 replied:
Your current proposal is very close to just editing the location event, but you are using a reference relation instead of a replace relation. To me it looks like you are replacing the old location with a new location, so... why not use the replace relation? That way you are basically editing the location, which gives you compatibility with other clients for free (unless they don't support edits for location events). Downside is some clients may bog down if an event is edited a lot, but that could happen with m.reference as well.

Having a state event to point at various events is fine for me. If it is just an event pointing to some event ids, it doesn't leak much data and can be useful to index stuff. We use the same for pinning messages, etc. And if you don't want to, you can just edit location messages and users will have to scroll up to the message to find it instead of finding it in a convenient beacon list. (Or search by event type).

Was just using search by event type ever considered instead of using state events? It obviously breaks in Element Web without the E2EE search plugin, but it should work fine in native clients or unencrypted rooms.

andybalaam replied:
> Your current proposal is very close to just editing the location event, but you are using a reference relation instead of a replace relation.

I don't think so ... the references are not state events, but the initial event is a state event. The two events have different purposes: m.beacon_info (the state event) is a marker to say "I am sharing", and m.beacon is "I am here". They have different content.

> That way you are basically editing the location, which gives you compatibility with other clients for free

Which clients? Are you suggesting emitting and editing the same events emitted as part of MSC3488?

> Having a state event to point at various events is fine for me.

That is close to what I am proposing, except the state event is referred-to instead of being updated to be the referrer. A difference is that my proposal has one state event per sharing event, whereas I think you are suggesting one state event per person. The downside of one per person is that the event will need to be edited, and editing state events is not allowed. (See [MSC2676 - Edge cases](https://github.com/uhoreg/matrix-doc/blob/aggregations-edits/proposals/2676-message-editing.md#edge-cases): "What can we edit? Only non-state events for now.")

> Was just using search by event type ever considered instead of using state events?

Searching by state events could work, but it would require whoever is doing the searching to have all timeline events available. The purpose of room state is to hold this kind of always-available information, so it seems the right place to do it.

deepbluev7 replied:
> I don't think so ... the references are not state events, but the initial event is a state event. The two events have different purposes: m.beacon_info (the state event) is a marker to say "I am sharing", and m.beacon is "I am here". They have different content.

Your "I am sharing event" could have a list of event ids of the events being edited. So if you start sharing a location, you send a location, then put the event id of that in the "I am sharing event" and just edit the location and when you stop, you remove the event id from the "I am sharing" event.

> Which clients? Are you suggesting emitting and editing the same events emitted as part of [MSC3488](https://github.com/matrix-org/matrix-doc/pull/3488)?

Exactly. But since you are using extensible events, editing your beacon event will fall back to being rendered as a location event and if that edits, you basically get an updating location in the client. It's nicely reusing existing functionality to build new functionality instead of reinventing everything.

> That is close to what I am proposing, except the state event is referred-to instead of being updated to be the referrer. A difference is that my proposal has one state event per sharing event, whereas I think you are suggesting one state event per person. The downside of one per person is that the event will need to be edited, and editing state events is not allowed. (See [MSC2676 - Edge cases](https://github.com/uhoreg/matrix-doc/blob/aggregations-edits/proposals/2676-message-editing.md#edge-cases): "What can we edit? Only non-state events for now.")

It doesn't really matter if you have multiple beacon infos or just one. You are just not supposed to use relations to edit state events, since state events are already edited by just sending a newer state event.

So basically my proposal was (with all the extra fluff omitted, that is the same as your proposal):

```json5
{
  type: beacon_info,
  live: ["$eventid"] // or just the event id and not an array, if you only want one location
}
```

```json5
{
  type: beacon,
  m.location: { /* ... */ },
  event_id: "$eventid"
}
```

And to update that just send a new beacon with a replace relation to "$eventid".

It is not much different from your proposal, but:

- Changing locations can be displayed in clients that just support edits and (extensible) location events without implementing this MSC
- You can still easily get the event from the beacon. Might be harder to get the beacon from the event, could be solved by using the eventid as the statekey, I guess.
- You can use the aggregation APIs to get the latest event more easily, no need to iterate over all references
- You are using "editing" the location. This uses the proper MSC for edits instead of reinventing it in a weird manner again (we already have edits for state events and we abuse reference relations for verifications, but at least in the second place it is because we send that as different users)

Anyway, your MSC looks fine, it just also looks like edits to me :D

andybalaam replied:
This makes a lot of sense, but editing previous locations seems odd for a use case that explicitly wants to preserve history. Of course edits don't delete history (or might not?) but semantically they are about changing an existing message, whereas with history I see it as making a list of previous locations at specific times.

It's neat that using edits and re-using the `m.location` type would potentially work with clients that don't support live shares, but I do argue the concepts don't quite fit together.

deepbluev7 replied:
I'm not quite clear what the difference is. In your case you are replacing the current instance of the location, but you want to have the history of it available. Messages do the same, you replace the current instance of it, but you can show the whole edit history. You could pretty much use the same code for the history stuff, while now you need to reimplement it slightly differently. There are some benefits to having the relation to the state event, but it still doesn't feel right, that edits are not composable enough to cover that use case.

andybalaam replied:
It's just the meaning: we're not editing a previous event - we're posting a new one for a new time.

andybalaam replied:
Thank you for your work on this. I wrote it up in 99861a90 - please comment if I didn't cover it properly.

deepbluev7 replied:
Looks good to me, thank you! <3

deepbluev7 commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+building on MSC3488 (m.location: Extending events with location data).
+
+We introduce two types of state events to describe beacons: the first contains
+the metadata about the beacons advertised by a given user: `m.beacon_info.*`.
```
Having variable event types has been an issue in the past for some implementations. There have been quite some complaints about those keys in account data and they make it very hard to list all state events of a specific type currently.

andybalaam replied:
I updated the proposal today to remove variable event types.

neilalexander commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+ensures that the current location of a given beacon is trivial to determine
+(rather than the client having to back-paginate history).  It also gives us
+location history where desirable.  If history is not wanted, then one can use
+data retention policies(e.g. exploding messages) to ensure it does not
```
Given that there isn’t really anything that guarantees that another server will have retention policies or obey redactions, I would be worried about the permanence of location information being stored in the timeline (state events or otherwise) in any form.

I wouldn’t expect if I shared my location through iMessage or whatever-other-product that my location data would accumulate endlessly in such a way that the other user could replay it later in days, weeks or even years to come. I would also expect that if I stopped sharing my location, that it would no longer be available for the user to download from the server in any form.

If a user forgets about either present or historic location sharing, there could be very real safety implications (i.e. abusive partners accessing your devices/chats, “friends” that may actually be stalking you etc).

Sending data ephemerally (or using send-to-device messaging) does design out historical use-cases but it also designs out a whole class of real-world safety issues.

deepbluev7 replied:
I feel like having this E2EE at least somewhat limits the impact, but E2EE for state events is still very far in the future and unencrypted state events basically give a server op an almost limitless amount of location data. That feels wrong. If this isn't at least ephemeral, it should in any case be encrypted. Even if it is ephemeral, it should be encrypted, because in theory ephemeral events can also just be stored indefinitely (but not shared to new servers).

neilalexander replied:
Encryption helps to stop some people you *didn’t* intend on seeing your location (i.e. homeserver administrators) but it doesn’t stop the problem of someone who physically gains access to your device being able to pull a potentially lengthy history later, nor does it provide any protection whatsoever if the encryption keys are compromised later on. When it comes to personal safety, the best data is no data at all — ephemeral sharing is much less likely to leave a footprint that outlives the user’s expectations.

deepbluev7 replied:
Sure, I just mean that ephemeral data would need to be encrypted too, since otherwise a lot of people would have access to your location instead of just your sharing target. You would get weekly complaints about "matrix.org has access to all your location data". For sensitive data, like real time location of people, encryption shouldn't be optional, no matter how it is sent.

neilalexander replied:
Agreed!

Sorunome replied:
encrypting EDUs should be doable exactly the same way as normal timeline PDUs, once we have custom EDUS.

ara4n replied:
So, to be clear: the rationale for supporting persisting location from the outset is to support the use cases of:

 * Publishing things like Strava-style exercise data
 * Bridging to things like ADS-B sources for sharing plane flight path data.
 * Logistics use cases where you want to track where your fleet of vehicles/ships are, or the packages within them
 * Supporting scenarios where you're trying to rendezvous with someone/something which will be aided by seeing their historical whereabouts (e.g. search-and-rescue operations; enhanced Find-my-iPhone style use cases).

Now, these are obviously different to the use case for publishing ephemeral location data (a la 'share my location' in WhatsApp or iMessage), which is typically "i want you to rendezvous with me, but you should be able to figure it out without seeing any historical data".

My reasons for combining the two in this MSC are:
 * If we don't, clients will need to implement two entirely different ways of handling live v. historical data - e.g. handling EDUs as well as timeline events.  This increases complexity and risk of partial implementations.
 * We **HAVE** to add burn-after-reading messages in the near future (MSC #2228 or similar) to maintain feature parity with other messengers (WhatsApp, Signal, Telegram, FB Messenger, SnapChat, etc. etc.).  The expectation is that, like redactions, there will be enormous social pressure on servers and clients to honour them, and in practice they will be honoured.  It would be horrifying to send a self-destructing message (be that IM or location data) to someone and discover their client or server was maliciously persisting it anyway - and it would just as horrifying to discover that you were sending them EDUs of ephemeral data and they were persisting that too.  Therefore, **if and only if we have self-destructing messages**, i am unconvinced that there is any additional privacy benefit to sending via EDUs versus self-destructing timeline events - and it will just increase implementation complexity and fragmentation risk in the long run.

So, perhaps the solution to this will depend on how timings for this landing compare with timings for rolling out self-destructing messages across the network.  If the self-destructing messages is slow to roll out, then I totally agree that we should split the MSC into separate 'live share' and 'historical share' MSCs, and ship 'live share' with hypothetical encrypted-per-room-EDUs as the transport - which will of course block this MSC on defining custom-per-room-EDUs.

On the other hand, if we have widespread uptake of self-destructing messages, I think we should kill two birds with one stone and make the most of it as a capability.

Sorunome replied:
> On the other hand, if we have widespread uptake of self-destructing messages, I think we should kill two birds with one stone and make the most of it as a capability.

Soru thinks this would be a rather bad idea as PDUs take up significantly more resources long-term due to *always* hanging around in the timeline, even when redacted.

Custom EDUs are needed badly anyways, as matrix wants to be more than just an IM protocol. Custom EDUs are needed for lots of IM purposes, so using them for live location sharing is, well, the next logical thing.

andybalaam replied:
This proposal has been updated to be only about use cases where we explicitly want the historical locations to be in the timeline. The use of EDUs has been split into https://github.com/matrix-org/matrix-spec-proposals/pull/3672

Half-Shot commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+data via to-device messages to interested devices.
+ * Pros:
+   * Doesn't thrash state resolution by storing new location points every time they change
+   * Gets you easy E2EE automatically
```
Might have missed this, but not sure how to-device messages provides E2EE. Unless you are encrypting the to-device messages themselves, the HS can presumably snoop on the contents of those messages?

deepbluev7 replied:
I think this means, that there is a specced way for how to encrypt to_device messages (similar to normal messages), but none for EDUs or state events.

andybalaam replied:
EDU-related stuff is now all in https://github.com/matrix-org/matrix-spec-proposals/pull/3672

Half-Shot commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,138 @@
+
+Alternatively, we could negotiate a WebRTC data channel using MSC3401 and
+stream low-latency geospatial data over the participating devices in the room.
+This doesn't give us historical data, however, and requiring a WebRTC stack
```
I am wondering if WebRTC really *is* that prohibitive. It's fairly easy to get going in Node / Go, and I presume there are bindings for other languages.

Given we depend on it for m.call already, it doesn't feel like they much of a stretch to say if you want real time locations of users then you need to be able do handle WRTC data channels.

Given all the privacy / performance concerns of the other solutions, this feels like the nicest channel to use to avoid those problems.

andybalaam replied:
I think WebRTC is an alternative to using EDUs, so I think this conversation needs to carry on in https://github.com/matrix-org/matrix-spec-proposals/pull/3672

stefanceriu committed:
Location streaming tweaks (#3622)

* Various tweaks following initial implementation:
- switched from `m.beacon.*` state event to `m.becon` reference relationship to original `m.beacon_info.*`
- changed `beacon_info` created to `m.ts`
- renamed `lifetime` to `timeout`
- added `m.asset` `type`

ara4n committed:
split out ephemeral location share to MSC3672

stefanceriu committed:
Add persitent location data use cases.

richvdh commented on proposals/3489-location-streaming.md:
````diff
@@ -0,0 +1,167 @@
+
+```json5
+{
+    "type": "m.beacon_info.@matthew:matrix.org",
````
this feels like a bit of a confusing example. Should it be something like

```suggestion
    "type": "m.beacon_info.beacon_12345",
```

or something?

ara4n replied:
well, it needs a unique ID of some kind, and given this example is showing me live-sharing my location, i suggested using my mxid given it's already a unique ID for describing me.  I guess it could be `m.beacon_info.1` or something instead, given the IDs are namespaced to me.

andybalaam replied:
Since the reason we are appending to the type is to allow multiple beacons per user (if I understand right), then a number seems better than using our mxid.

andybalaam replied:
Updated today to avoid variable event types.

Johennes commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,167 @@
+by letting the final part of the event type be a unique ID for the beacon.
+(Ideally we would put the unique ID as its `state_key` and flip a bit on the
+event to indicate that it can only ever be written by its owner - we may get
+this via MSC3414 encrypted state events).
```
Continuing a conversation that happened elsewhere here.

@ara4n said:

> on variable event types: i’d failed to think through the permissions problem [...]
> so my suggestion would be to switch to a static event type, and in parallel we do an msc to flip a bit on the event to state “this can only by set by the author”

andybalaam replied:
Updated today to avoid variable event types.

andybalaam committed:
Remove variable event types; clarifications

* Clarify that this proposal includes history - see MSC3672 for EDUs
* Note dependency on MSC3267 for reference relationships
* Remove variable event_type - instead use state_key
* Propose state_key prefixes inspired by MSC3757 and MSC3779
* Update links to matrix-spec-proposals
* Clarify how to stop sharing, and when a share is stopped
* Mention the need for redactions to work
* Clarify the use of milliseconds for all timing
* Mention implementations

andybalaam committed:
Fix spelling error

andybalaam committed:
Add a reference to MSC2244: Mass redactions

uhoreg commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,462 @@
+
+If `state_key` exactly equals the sender's mxid, then the current Matrix spec
+prevents anyone else from overwriting the state from this event. If `state_key`
+contains further characters, that protection will not apply until
+[MSC3757](https://github.com/matrix-org/matrix-spec-proposals/pull/3757) (or
+some equivalent mechanism) is available.
```
I think this is worded incorrectly or confusingly.  Without MSC3757, nobody will be allowed to set an event with a `state_key` of the form `[userId]_[anything]`.  So it isn't that the "protection will not apply", it is simply not possible at all.

andybalaam replied:
Thank you - corrected in multiple places.

uhoreg commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,462 @@
+We could include velocity data as well as displacement data here (especially as
+the W3C geolocation API provides it); this has been left for a future MSC in
+the interests of avoiding scope creep.
+
```
Another possibility for supporting multiple beacons, which wouldn't require a new room version, is to make `m.beacon_info` into a set of beacons, rather than a single beacon.  e.g.

```json5
{
  "type": "m.beacon_info",
  "state_key": "@hubert:uhoreg.ca",
  "content": {
    "some_id": {
      // beacon description ...
    },
    "some_other_id": {
      // another beacon description
    }
  }
}
```

So if a client wants to add a new beacon, it would just add a new entry to the user's `m.beacon_info` state event.  The `m.beacon` events would then somehow reference the beacon IDs, rather than the beacon event ID.

What if multiple clients want to add a new beacon at the same time?  The beacon info can also be stored in the user's per-room account data, under the event type `m.beacon_info.[id]`.  Since it's the user's own data, there is no worry about flooding anyone else.  All the clients can then work together to ensure that the `m.beacon_info` state event is synchronized with the `m.beacon_info.*` account data.  e.g. when a client notices that the account data has been changed, it checks the room state to see if the state matches the account data.  If not, then it updates the room state (probably after a random delay, so that multiple clients won't try to set the room state at exactly the same time).

andybalaam replied:
Hopefully covered as an alternative in 99861a90 - do comment if I didn't cover it.

andybalaam committed:
Re-word to reflect that mxid_blah is not allowed as a state_key

andybalaam committed:
Write up state-event-per-user idea

HarHarLinks commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+events will automatically have permission to share their live location, because
+the `state_key` matches the pattern required to be an "owned" state event.
+
+## Alternatives
```
There is another alternative that appears not to have been discussed yet, although the idea is quite obvious:

Instead of introducing a new event `m.beacon` that for the most part duplicates an MSC3488 `m.location`:
- use `m.location`
- when used for location streams (i.e. in this context and MSC3672), require the `m.relates_to` mixin to be set

Pros compared to the MSC as it currently is:
- automatic fallback for clients that support static `m.location`
- further fallback to `m.text`

Cons:
- this could result in clients without live location support having their timelines spammed. while one could still view the movement trail similar to the original intention, this might be considered undesireable

zecakeh commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+`description` is the same as an `m.location` description
+([MSC3488](https://github.com/matrix-org/matrix-spec-proposals/pull/3488)).
+
+`live` should be true when a user starts sharing location.
```
When `live` is false, `m.ts` and `timeout` become meaningless, right? So those fields are optional?

If they are, couldn’t we get rid of `live` and say that the location sharing is live when both `m.ts` and `timeout` are present and the timeout is not reached?

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+
+We introduce two types of event to describe beacons:
+
+* **`m.beacon_info`**: a single state event containing metadata about this
```
It'd be useful if the initial `beacon_info`  message included the first `m.beacon` data packet. On the client side, we need a "pending location" UI state in case there's large latency between these two messages being received, parsed, and rendered. We could avoid that entirely if the initial message included the first data point.

Is there any reason why this couldn't always be true? Would there ever be a reason to send the first event where you didn't know the data that would be in the first message already? I can't think of one at least.

Johennes replied:
I think this would technically be possible but the downside is that the `beacon_info` event is a state event and state events are not encrypted. So you'd disclose your initial location to the homeserver this way.

Susurrus replied:
Forgot about that, good point. Could you embed the position as an encrypted payload within the state event?

Johennes replied:
I suppose yes but then you'd still have to share the key for it. It might be better to wait for https://github.com/matrix-org/matrix-spec-proposals/pull/3414 (or something similar).

Susurrus replied:
I've realized we can just wait until a valid location has been received to render the timeline item. I see little downside to rendering it this way, and a big advantage as it reduces UI state that needs to be tracked. I'm going to try implementing this way, and if this does seem a reasonable path, might be worth mentioning it explicitly here as a MAY or SHOULD.

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+
+Otherwise it is not live.
+
+#### How to display a single location share event
```
Should there be any additional guidance for when there are multiple location shares happening simultaneously? We've been trying to figure out how we should render this. Element clients show multiple map timeline views, but allow you to see all shares within the same map widget. We're wondering if that's the intention with this sharing. There doesn't seem to be a way to specify in these messages a way to intentionally group assets. I would think having some grouping would make sense, but I'm not certain the API for it.

For example, with the static location sharing, you might want all assets to show on a single map because they're all related. For live locatoin, you might want to combine all `m.self` shares by default, because everyone's doing the same thing, but maybe there's a second group of people who want to share their location as a group.

Until a clear use case emerges, maybe recommending showing all `m.self` locations on the same map would be good.

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+`m.beacon` events should be sent at a variable rate which is meaningful for
+the use case for the asset being tracked. This rate should not be more
+frequent than sending an event every 2 seconds. If the asset is not moving,
+the location should still be refreshed on a regular basis to convey that
```
Should this be written with the more formal "SHOULD" here?

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+the use case for the asset being tracked. This rate should not be more
+frequent than sending an event every 2 seconds. If the asset is not moving,
+the location should still be refreshed on a regular basis to convey that
+information - e.g. every 30 seconds.
```
Should this 30s be a default recommendation for clients? It's mentioned off-hand here, but using SHOULD here would make that more clear.

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+* `m.ts + timeout < now_ts` where `now_ts` is the current time in milliseconds
+  since the epoch.
+
+Otherwise it is not live.
```
"dead" is a more common way to refer to this. I've been running into this minor naming convention gap when coding this up in Fractal: to set something as dead, I've been using "inactive" instead to avoid using "dead" cause "set_not_live" reads awkwardly.

Maybe we could change the "live" nomenclature entirely to "active" & "inactive"?

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+`m.asset` identifies the type of asset being tracked as per
+[MSC3488](https://github.com/matrix-org/matrix-spec-proposals/pull/3488).
+
+#### `state_key` of `m.beacon_info`
```
Is this a required field? Or could it be inferred to be the sender's mxid if not included? This would make the simple case of sharing user location a bit more succinct. But couldn't we get the same behavior by namespacing the state_key to the sender? Or is there a reason you'd want the name to referenced by multiple senders? But you could instead prepend the sender's mxid if one isn't provided. This implictly works when the field is missing entirely too then, as the sender's mxid + null is just their mxid.

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+`m.ts` is the time when location sharing started, in milliseconds since the
+epoch.
+
+`timeout` is the length of time in milliseconds that the location will be live.
```
Is millisecond precision required here? Since this is a timeout, second precision seems sufficient.

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+  *Extending events with location data*. This MSC re-uses the `m.location`
+  object from MSC3488.
+
+### MSCs that will enhance this one
```
Both of these have been closed in favor of https://github.com/matrix-org/matrix-spec-proposals/pull/4354.

I was trying to figure out what permissions in a room this currently requires as I ran into that problem. Until Sticky Events lands, would be good to explicitly state what permissions are being reused for this. Cause in EW I don't see any permission that could be applicable

Susurrus commented on proposals/3489-location-streaming.md:
```diff
@@ -0,0 +1,483 @@
+For clients with timelines, we expect that a map or similar will appear in the
+timeline whenever a `m.beacon_info` event with `live: true` is sent, and
+related `m.beacon` events will cause that map to update. Stopping sharing
+(sending a `m.beacon_info` event with `live: false`) might not create any new
```
Should read receipts be sent for these messages as normal or could they be ignored for this? Since they won't be shown, as the start event tracks that, is it needed?
